### PR TITLE
refactor(crypto): minor refactor of decryptWithAttachments

### PR DESCRIPTION
**Pull request by seaerchin**
## Problem 
Refactored `decryptWithAttachments` for additional clarity

## Issues faced
1. lack of chaining means that everything has to be within a `try/catch` block

### CODE REVIEW

File: src/util/crypto.ts [hunk 1]
```diff
@@ -79,3 +79,16 @@ export const verifySignedMessage = (
...
   return JSON.parse(encodeUTF8(openedMessage))
 }
+
+/**
+ * Helper method to check  if all the fields given are wtihin the filenames
+ * @param fieldIds the list of fieldIds to check
+ * @param filenames the filenames that should contain the fields
+ * @returns boolean indicating whether the fields are valid
+ */
+export const areFieldsValid = (
```
Comment (mantariksh) [inline src/util/crypto.ts]: can we name this something more specific, e.g. `areAttachmentFieldIdsValid`?

File: src/util/crypto.ts [hunk 2]
```diff
@@ -79,3 +79,16 @@ export const verifySignedMessage = (
     throw new Error('Failed to open signed message with given public key')
   return JSON.parse(encodeUTF8(openedMessage))
 }
+
+/**
+ * Helper method to check  if all the fields given are wtihin the filenames
```
Comment (mantariksh) [inline src/util/crypto.ts]:
```suggestion
 * Helper method to check if all the field IDs given are within the filenames
```

File: src/crypto.ts [hunk 3]
```diff
@@ -206,7 +207,8 @@ export default class Crypto {
    * @param formSecretKey Secret key as a base-64 string
    * @param decryptParams The params containing encrypted content and information.
    * @returns A promise of the decrypted submission, including attachments (if any). Or else returns null if a decryption error decrypting any part of the submission.
-   * @throws {MissingPublicKeyError} if a public key is not provided when instantiating this class and is needed for verifying signed content.
+   * @throws {MissingPublicKeyError} if a public key is not provided when instantiating this class and is needed for verifying signed content.]
+   * @throws {MissingFilenameError} if a file had an invalid name or if the name was not found
```
Comment (mantariksh) [inline src/crypto.ts]: this is not a thing right?

File: src/crypto.ts [hunk 4]
```diff
@@ -226,37 +229,45 @@ export default class Crypto {
...
+            })
+            .then((decryptedFile) => {
+              if (decryptedFile === null)
+                throw new Error('Attachment decryption failed')
+              decryptedRecords[fieldId] = {
+                filename: filenames[fieldId],
+                content: decryptedFile,
+              }
+            })
+        })
```
Comment (mantariksh) [inline src/crypto.ts]: I actually prefer the old method of assigning the Promise array to a variable first to avoid so much nesting, wdyt?

File: src/util/crypto.ts (hunk 1 above)
Comment (seaerchin) [reply src/util/crypto.ts]: you're right and i'll update accordingly!

File: src/crypto.ts (hunk 3 above)
Comment (seaerchin) [reply src/crypto.ts]: yeap, oversight on my end. will remove!

File: src/util/crypto.ts (hunk 2 above)
Comment (seaerchin) [reply src/util/crypto.ts]: it's time for me to install a spell-checker @__@

File: src/crypto.ts (hunk 4 above)
Comment (seaerchin) [reply src/crypto.ts]: i don't have a strong opinion on this; main conflict here seems to be nesting vs mutating - i just don't like mutating HAHA but yea, will change accordingly!

File: src/crypto.ts [hunk 5]
```diff
@@ -226,37 +228,49 @@ export default class Crypto {
...
+              submissionPublicKey:
+                encryptedAttachment.encryptedFile.submissionPublicKey,
+              nonce: encryptedAttachment.encryptedFile.nonce,
+              binary: decodeBase64(encryptedAttachment.encryptedFile.binary),
+            }
+            return this.decryptFile(formSecretKey, encryptedFile)
+          })
+          .then((decryptedFile) => {
+            if (decryptedFile === null)
+              throw new Error('Attachment decryption failed')
```
Comment (liangyuanruo) [inline src/crypto.ts]: can introduce a custom error?

File: src/crypto.ts [hunk 6]
```diff
@@ -226,37 +228,49 @@ export default class Crypto {
...
+          .then((downloadResponse) => {
+            const encryptedAttachment: EncryptedAttachmentContent =
+              downloadResponse.data
+            const encryptedFile: EncryptedFileContent = {
+              submissionPublicKey:
+                encryptedAttachment.encryptedFile.submissionPublicKey,
+              nonce: encryptedAttachment.encryptedFile.nonce,
+              binary: decodeBase64(encryptedAttachment.encryptedFile.binary),
+            }
+            return this.decryptFile(formSecretKey, encryptedFile)
```
Comment (liangyuanruo) [inline src/crypto.ts]: suggest extracting this section into a function to be even cleaner

Comment (liangyuanruo) [reply src/crypto.ts]: then you could even use a ternary op to decide whether to throw the error in the next part instead of chaining another `then` block

File: src/crypto.ts [hunk 7]
```diff
@@ -226,37 +228,49 @@ export default class Crypto {
...
-        }
-
-        return this.decryptFile(formSecretKey, encryptedFile)
-      }).then((decryptedFile) => {
-        if (decryptedFile === null) throw new Error("Attachment decryption failed")
-        decryptedRecords[fieldId] = { filename: filenames[fieldId], content: decryptedFile }
-      }))
+    const fieldIds = Object.keys(attachmentRecords)
+    // Check if all fieldIds are within filenames
+    if (!areAttachmentFieldIdsValid(fieldIds, filenames)) {
```
Comment (liangyuanruo) [inline src/crypto.ts:234]: this seems important - maybe we could add unit tests to make sure it is called + behaves correctly?

Comment (mantariksh) [reply src/crypto.ts:234]: there's an existing test for this: https://github.com/opengovsg/formsg-javascript-sdk/blob/8dd992272e1520b6d09b7d1b23b932863fde5fd7/spec/crypto.spec.ts#L350

File: src/crypto.ts (hunk 5 above)
Comment (mantariksh) [reply src/crypto.ts]: this is meant to be caught internally right? regardless of what kind of error is thrown, we return `null`

Comment (seaerchin) [reply src/crypto.ts]: i wrote a custom error at the start but i removed it because my understanding is that our library treats `errors` not as values but as exceptions, and in that sense, should never throw (as much as possible). 

will change if needed!

File: src/crypto.ts (hunk 6 above)
Comment (seaerchin) [reply src/crypto.ts]: good point, will do so!

File: src/crypto.ts (hunk 7 above)
Comment (seaerchin) [reply src/crypto.ts:234]: could i clarify if this means that the helper should be unit tested? there's an existing test which @mantariksh linked which checks this case

Comment (liangyuanruo) [reply src/crypto.ts:234]: as discussed, to add a small unit test for `areAttachmentFieldIdsValid`

**Review by liangyuanruo — APPROVED** [review]

File: spec/util.spec.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import { areAttachmentFieldIdsValid } from '../src/util/crypto'
+describe('utils', () => {
+  describe('areAttachmentFieldIdsValid', () => {
+    it('should return true when all the fieldIds are within the filenames', () => {
+      // Arrange
+      const MOCK_FILENAMES = {
+        mock: 'file',
+        alsomock: 'file2',
+      }
+      const MOCK_FIELD_IDS = ['mock', 'alsomock']
```
Comment (mantariksh) [inline spec/util.spec.ts]: you can use `Object.keys(MOCK_FILENAMES)` to make this more generalisable

File: spec/util.spec.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      expect(actual).toBe(true)
+    })
+
+    it('should return false when some fieldIds are not within the filenames', () => {
+      // Arrange
+      const MOCK_FILENAMES = {
+        mock: 'file',
+        alsomock: 'file2',
+      }
+      const MOCK_FIELD_IDS = ['mock', 'missingField']
```
Comment (mantariksh) [inline spec/util.spec.ts]: or `Object.keys(MOCK_FILENAMES).concat('missingField')`

File: src/crypto.ts [hunk 10]
```diff
@@ -226,37 +229,51 @@ export default class Crypto {
...
-        }
+    const downloadPromises = fieldIds.map((fieldId) => {
+      return (
+        axios
+          // Retrieve all the attachments as JSON
+          .get<EncryptedAttachmentContent>(attachmentRecords[fieldId], {
+            responseType: 'json',
+          })
+          // Decrypt all the attachments
+          .then(async ({ data: downloadResponse }) => {
```
Comment (mantariksh) [inline src/crypto.ts]: don't mix `async`-`await` and `Promise.then`. this works, but [mixing styles can be confusing](https://stackoverflow.com/questions/54385676/whats-wrong-with-awaiting-a-promise-chain/54387912). chain the Promises instead.

File: src/errors.ts [hunk 11]
```diff
@@ -33,8 +33,19 @@ class WebhookAuthenticateError extends Error {
   }
 }
 
+class AttachmentDecryptionError extends Error {
+  constructor(message = 'Attachment decryption with the given nonce failed.') {
+    super(message)
+    this.name = this.constructor.name
+    // Set the prototype explicitly.
+    // See https://github.com/facebook/jest/issues/8279
+    Object.setPrototypeOf(this, AttachmentDecryptionError.prototype)
```
Comment (mantariksh) [inline src/errors.ts]: I don't think you need these because this error is never tested for in Jest (and can't be tested since it's meant to be caught within the same function)

**Review by mantariksh — CHANGES_REQUESTED** [review]

File: src/errors.ts [hunk 12]
```diff
@@ -33,8 +33,16 @@ class WebhookAuthenticateError extends Error {
   }
 }
 
+class AttachmentDecryptionError extends Error {
+  constructor(message = 'Attachment decryption with the given nonce failed.') {
+    super(message)
+    this.name = this.constructor.name
```
Comment (mantariksh) [inline src/errors.ts:39]: out of curiosity, why is this line necessary?

**Review by mantariksh — APPROVED** [review]

File: src/errors.ts (hunk 12 above)
Comment (seaerchin) [reply src/errors.ts:39]: it isn't! i actually left it in because i thought it was convention @__@ i'll remove it!

Comment (seaerchin) [reply src/errors.ts:39]: i went to investigate more and it was q enlightening for me! the reason why this is done is because it's an error. when js (and by extension, ts) throws an error, the error printed to the output is done using  `error.name + ' ' + error.message` - this means that when the error here is thrown, because it's missing the name property, it will default to `Error` (as it calls `super(message)`). While the argument might be that it's guaranteed to never be thrown, i think this is a relatively small change that helps in the future so i'd just leave it in!
